egs2/wsj0-2mix/enh1/enh.sh

```diff
@@ -0,0 +1,682 @@
+        _train_data_param+="--train_data_path_and_name_and_type ${_enh_train_dir}/dereverb.scp,dereverb_ref,sound "
+        _train_shape_param+="--train_shape_file ${enh_stats_dir}/train/dereverb_ref_shape "
+        _valid_data_param+="--valid_data_path_and_name_and_type ${_enh_dev_dir}/dereverb.scp,dereverb_ref,sound "
+        _valid_shape_param+="--valid_shape_file ${enh_stats_dir}/valid/dereverb_ref_shape "
+        _fold_length_param+="--fold_length ${_fold_length} "
+    fi
+
+    if $use_noise_ref; then
+        # reference for denoising
+        for n in $(seq "${noise_type_num}"); do
+            _train_data_param+="--train_data_path_and_name_and_type ${_enh_train_dir}/noise${n}.scp,noise_ref${n},sound "
+            _train_shape_param+="--train_shape_file ${enh_stats_dir}/train/noise_ref${n}_shape "
+            _valid_data_param+="--valid_data_path_and_name_and_type ${_enh_dev_dir}/noise${n}.scp,noise_ref${n},sound "
+            _valid_shape_param+="--valid_shape_file ${enh_stats_dir}/valid/noise_ref${n}_shape "
+            _fold_length_param+="--fold_length ${_fold_length} "
+        done
+    fi
+
+
+    log "enh training started... log: '${enh_exp}/train.log'"
+    # shellcheck disable=SC2086
+    python3 -m espnet2.bin.launch \
+        --cmd "${cuda_cmd} --name ${enh_exp}/train.log --mem ${mem}" \
+        --log "${enh_exp}"/train.log \
+        --ngpu "${ngpu}" \
+        --num_nodes "${num_nodes}" \
+        --init_file_prefix "${enh_exp}"/.dist_init_ \
+        --multiprocessing_distributed true -- \
+        python3 -m espnet2.bin.enh_train \
+            ${_train_data_param} \
+            ${_valid_data_param} \
+            ${_train_shape_param} \
+            ${_valid_shape_param} \
+            ${_fold_length_param} \
+            --resume true \
+            --output_dir "${enh_exp}" \
+            ${_opts} ${enh_args}
+
+fi
+
+
+if [ ${stage} -le 7 ] && [ ${stop_stage} -ge 7 ]; then
+    log "Stage 7: Enhance Speech: training_dir=${enh_exp}"
+
+    if ${gpu_inference}; then
+        _cmd=${cuda_cmd}
+        _ngpu=1
+    else
+        _cmd=${decode_cmd}
+        _ngpu=0
+    fi
+
+    _opts=
+
+    for dset in "${dev_set}" ${eval_sets}; do
+        _data="${data_feats}/${dset}"
+        _dir="${enh_exp}/separate_${dset}"
+        _logdir="${_dir}/logdir"
+        mkdir -p "${_logdir}"
+
+        _scp=wav.scp
+        _type=sound
+
+        # 1. Split the key file
+        key_file=${_data}/${_scp}
+        split_scps=""
+        _nj=$(min "${infernece_nj}" "$(<${key_file} wc -l)")
+        for n in $(seq "${_nj}"); do
+            split_scps+=" ${_logdir}/keys.${n}.scp"
+        done
+        # shellcheck disable=SC2086
+        utils/split_scp.pl "${key_file}" ${split_scps}
+
+        # 2. Submit decoding jobs
+        log "Separation started... log: '${_logdir}/enh_inference.*.log'"
+        # shellcheck disable=SC2086
+        ${_cmd} --gpu "${_ngpu}" JOB=1:"${_nj}" "${_logdir}"/enh_inference.JOB.log \
```

Comment:
We may prepare different GPU variables for separation (e.g., see https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/asr.sh#L31-L32 in ASR recipes). We may perform CPU inference for separation.

Reply:
This was already done. Please refer to the code below: https://github.com/espnet/espnet/blob/b48d6393f3436ee174dc8739f4099550c90290db/egs2/wsj0-2mix/enh1/enh.sh#L536-L542